Omit address release if the address still responds to ping

jnummelin opened the pull request:
Workaround for the different cases of leaking network namespaces. The cleanup routine will eventually release addresses that stop responding to pings.

jakolehm commented on lib/ipaddr_helpers.rb:
```diff
@@ -244,4 +245,9 @@ def supernets
       end
     end
+
+    def ping?
```
I'm against this.. we should not add new methods to monkey patched IPAddr unless really necessary.

jnummelin replied:
Roger. I'll move this out.

jakolehm commented on lib/ipaddr_helpers.rb:
```diff
@@ -1,4 +1,5 @@
 require 'ipaddr'
+require 'net/ping'
```
Probably we need only `net/ping/icmp`?

jnummelin replied:
yes

SpComb commented on app/mutations/addresses/release.rb:
```diff
@@ -31,10 +32,18 @@ def execute
-        info "Delete address=#{address.id} in pool=#{@pool.id}"
-
-        address.delete!
+        if ping?(address.address)
```
With the way this is written as a flat if-elsif chain, this should be `elsif`.

SpComb commented on app/mutations/addresses/release.rb:
```diff
@@ -31,10 +32,18 @@ def execute
       end
     end
+
+    def ping?(ip_address)
```
```ruby
@raise [RuntimeError] requires root privileges
@return [Boolean]
```

SpComb commented on app/mutations/addresses/release.rb:
```diff
@@ -31,10 +32,18 @@ def execute
-
-        address.delete!
+        if ping?(address.address)
+          info "#{self.address} still responds to ping, deleting omitted"
```
This should be a `warn`

```ruby
warn "Skip zombie address=#{address.id} in pool=#{@pool.id} that still responds to ping"
```

SpComb replied:
And this should also add an error, so that the `Addresses::Release` fails. That way this will also show up in the agent logs, where it's easier to find.

jnummelin replied:
Is it really a good idea to make this error out the API call on agent side? Isn't this kind-of deferred release so from agent point-of view it would still mean successful API call?

SpComb replied:
It wasn't released. So I think it should be some kind of "4xx temporary error try again later" kind of case.

jnummelin replied:
would 202-Accepted work?
> 202 Accepted
> The request has been accepted for processing, but the processing has not been completed. The request might or might not be eventually acted upon, and may be disallowed when processing occurs.

IMO 4xx are "true" errors and this case is now something expected and not a real error :)

SpComb commented on app/mutations/addresses/release.rb:
```diff
@@ -31,10 +32,18 @@ def execute
     end
+
+    def ping?(ip_address)
+      icmp = Net::Ping::ICMP.new(ip_address.to_s)
```
I would pass an explicit `timeout` parameter to `Net::Ping::ICMP`, because this mutation is expected to timeout in all normal cases.

The default timeout seems to be `5`.

jnummelin replied:
good point. weave addresses are fast so I'll add 1 sec timeout on it

SpComb commented on app/mutations/addresses/release.rb:
```diff
@@ -31,18 +31,21 @@ def execute
+    #@return [Boolean]
     def ping?(ip_address)
-      icmp = Net::Ping::ICMP.new(ip_address.to_s)
+      icmp = Net::Ping::ICMP.new(ip_address.to_s, 0, 1)
```
I suggest something more self-documenting like

```ruby
def ping?(ip_address, timeout: 1.0)
  port = 0
  icmp = Net::Ping::ICMP.new(ip_address.to_s, port, timeout)
  ...
```

This `0, 1` part is way too hard to figure out when reading.

SpComb commented on plugin.rb:
```diff
@@ -108,12 +108,21 @@ def ensure_models
+      json({})
+    rescue Mutations::ValidationException => exc
+      if exc.errors['address'] && exc.errors['address'].symbolic == :zombie
+        status 202
```
I don't think this should be considered as a "normal" case. It should really be logged and noticed eventually, because it's useful as a sign of some new netns leak cropping up that really also needs to get fixed. We're hopefully mitigating the damage, but you can't ignore the problem completely.

I wouldn't make this a 2xx response, because the end result of this error case is exactly the same as if the `/IpamDriver.ReleaseAddress` request had not been done at all. The IPAM rejects the request and takes no action.

Agreed that it shouldn't necessarily be the generic "HTTP 400" that things like passing an invalid IP address would return. I would perhaps consider something like a HTTP 409 Conflict?

Re HTTP 202 specifically, that seems to be for something like the `kontena service deploy` case of asynchronous request processing:

> The representation sent with this response ought to describe the request's current status and point to (or embed) a status monitor that can provide the user with an estimate of when the request will be fulfilled.

jnummelin replied:
I guess this is one of the cases where none of the existing HTTP status codes really match well. :/ 409 kinda fits as in "Yes, the address should be released but cannot as it's in conflicting state (still responds to ping)".

I'm ok using 409. @jakolehm agree?

jakolehm replied:
My vote would go to 202 because address will be released at some point. Client can still log that situation by checking the status code, right?

jakolehm replied:
Actually just realized that cleanup is not "automatic", it's triggered by the agent. So probably 409 is more logical here.

SpComb commented:
Tested the `ping?` logic to work for a real zombie container:

```
W, [2017-03-28T15:32:29.843992 #1]  WARN -- Addresses::Release: Skip zombie address=10.81.128.10 in pool=kontena that still responds to ping
127.0.0.1 - - [28/Mar/2017:15:32:29 +0000] "POST /IpamDriver.ReleaseAddress HTTP/1.1" 202 2 0.0261
```

jnummelin commented:
409 got more "votes" so I changed the code to 409 in case of zombie release. 👻 

Added ping check also to cleanup mutation using `PingHelper` module.
